File: vibevm/vibepacks/org.vibevm.doc/web/v0.1.0/site/src/seo/resolve.ts
/** @scope spec://org.vibevm.core/vibevm/common/PROP-057#SEO-MANIFEST-AND-RESOLVER */

/**
 * The resolver on a static host — a redirect table and the
 * one page that reads it.
 *
 * A server would answer this with a 302 computed from the address map,
 * which is deterministic and needs no index (D-06). A static host cannot
 * answer anything, so F-15 says the public site publishes the map as a
 * table and a page that follows it in the browser. The local reader,
 * which does have a server, keeps the route.
 *
 * The table is written from the manifests — every document, under both
 * spellings of its version — and carries each page's anchors so the
 * resolver can tell a fragment that exists from one that does not, and
 * land the reader on the page rather than on nothing when a citation has
 * gone stale. It never invents: an address the table does not hold is
 * reported to the reader, not guessed at.
 */

import { docFileHref, doorHref, LATEST, type Address } from "./editions.ts";

/** Where the table and the page that reads it are served. */
export const RESOLVE_TABLE = docFileHref("resolve.json");
export const RESOLVER_PAGE = docFileHref("resolve/");

/** What the table says about one document. */
export type ResolveEntry = {
  /** The page's address on this site. */
  readonly href: string;
  /** Every named anchor the page carries, in document order. */
  readonly anchors: readonly string[];
};

/** The table itself, as `resolve.json` holds it. */
export type ResolveTable = {
  readonly schema_version: 1;
  /** Keyed by `spec://` address without a fragment. */
  readonly documents: Readonly<Record<string, ResolveEntry>>;
};

/**
 * The `spec://` address of a page, in the spelling the table is keyed
 * by: with the version for a numbered address, without it for `latest`.
 *
 * Both are real citations. A citation that names a version quotes that
 * contract; one that omits it means «whichever is newest», which is what
 * the `latest` address is (D-06). The language is in neither: a
 * `spec://` address names the documentation, and which language a reader
 * read it in is not part of what they cite.
 */
function keyOf(address: Address): string | null {
  if (address.kind !== "page" || address.document === undefined) return null;
  if (address.edition.segment !== null) return null;
  const card = address.edition.manifest.package;
  const coordinate =
    address.version === LATEST
      ? `${card.group}/${card.name}`
      : `${card.group}/${card.name}@${address.version}`;
  return `spec://${coordinate}/${address.document}`;
}

/** The redirect table for every address the source editions publish. */
export function resolveTableOf(addresses: readonly Address[]): ResolveTable {
  const documents: Record<string, ResolveEntry> = {};
  for (const address of addresses) {
    const key = keyOf(address);
    if (key === null) continue;
    documents[key] = {
      href: address.href,
      anchors: address.page?.anchors ?? [],
    };
  }
  return { schema_version: 1, documents };
}

/**
 * The script the resolver page runs, as one line.
 *
 * It is inline and in the head for the same reason the theme script is:
 * everything it does must happen before anything is painted, and a
 * reader who asked to be sent somewhere should not first watch a page
 * arrive. Its `sha256` is in the policy the build writes, computed from
 * these bytes (X-035).
 *
 * It reads three things and trusts none of them. The `uri` parameter is
 * matched against the shape of a `spec://` address before it is used as
 * a key; the destination comes from the table and is never built from
 * the parameter, so a crafted `uri` cannot name an address of its own;
 * and the fragment travels only when the page says it has that anchor.
 */
const RESOLVER_SCRIPT = [
  "(function(){",
  "var q=new URLSearchParams(location.search).get('uri');",
  "function fail(m){document.addEventListener('DOMContentLoaded',function(){",
  "var e=document.getElementById('answer');if(e)e.textContent=m;});}",
  "if(!q||!/^spec:\\/\\/[^\\s]+$/.test(q)){fail('No spec:// address was asked for.');return;}",
  "var hash=q.indexOf('#');",
  "var key=hash===-1?q:q.slice(0,hash);",
  "var anchor=hash===-1?'':q.slice(hash+1);",
  `fetch(${JSON.stringify(RESOLVE_TABLE)}).then(function(r){return r.json();}).then(function(t){`,
  "var e=t.documents[key];",
  "if(!e){fail('This site does not carry '+key+'.');return;}",
  "var at=e.href+(anchor&&e.anchors.indexOf(anchor)!==-1?'#'+anchor:'');",
  "location.replace(at);",
  "}).catch(function(){fail('The resolver table could not be read.');});",
  "})()",
].join("");

/**
 * The resolver page.
 *
 * Hand-written rather than a route of the application, and deliberately:
 * this address is a machine's entry point, its answer is a redirect, and
 * loading a framework to perform one is a second's wait paid by every
 * agent that follows a citation. It is `noindex` because it is not a
 * page — it is a junction — and it says in words what it is doing, so a
 * reader who lands here with scripting off is not left staring at a
 * blank document.
 */
export function resolverPage(): string {
  return [
    "<!DOCTYPE html>",
    '<html lang="en">',
    "<head>",
    '<meta charset="utf-8">',
    '<meta name="viewport" content="width=device-width, initial-scale=1">',
    '<meta name="robots" content="noindex, follow">',
    "<title>Resolve a spec:// address</title>",
    `<script>${RESOLVER_SCRIPT}</script>`,
    "</head>",
    "<body>",
    "<h1>Resolve a spec:// address</h1>",
    `<p>Ask for a page by its citation: <code>${RESOLVER_PAGE}?uri=spec://&lt;group&gt;/&lt;name&gt;/&lt;document&gt;#&lt;anchor&gt;</code>.`,
    " An address without a version resolves to the newest publication.</p>",
    '<p id="answer">Resolving&hellip;</p>',
    `<p><a href="${doorHref()}">The documentation this site carries</a>`,
    ` &middot; <a href="${docFileHref("manifest.json")}">manifest.json</a>`,
    ` &middot; <a href="${RESOLVE_TABLE}">resolve.json</a></p>`,
    "</body>",
    "</html>",
    "",
  ].join("\n");
}

/** The script the page carries, for the policy that has to name its hash. */
export const RESOLVER_INLINE_SCRIPT = RESOLVER_SCRIPT;

File: crates/vibe-doc-server/src/routes/resolve.rs
//! `<base>resolve?uri=spec://…` — a citation, followed (PROP-057
//! `##SEO-MANIFEST-AND-RESOLVER`; the plan's fork F-15).
//!
//! An agent quotes a rule by its `spec://` address, and a person handed
//! that address wants the page. On a static host there is no server to
//! ask, so the site publishes a redirect table and a page that follows it
//! in the browser; here there IS a server, and F-15 says the local reader
//! keeps the route. Same answer, computed instead of looked up.
//!
//! It invents nothing. The coordinate must be the one package this reader
//! was pointed at, the document must be a page that package actually
//! holds, and a fragment travels only when the page carries that anchor —
//! a citation that has gone stale lands the reader on the page rather
//! than on nothing, and an address this reader does not serve is reported
//! rather than guessed at.

specmark::scope!("spec://org.vibevm.core/vibevm/common/PROP-057#SEO-MANIFEST-AND-RESOLVER");

use axum::response::Response;

use crate::Reader;
use crate::error::ApiError;

/// The address the resolver answers at, under the reader's base.
pub(crate) const ROUTE: &str = "resolve";

/// One parsed citation.
#[derive(Debug, PartialEq, Eq)]
pub(crate) struct Citation {
    /// `<group>/<name>`.
    pub(crate) coordinate: String,
    /// The version the citation named, when it named one. Absent means
    /// «whichever is newest», which is what the `latest` address is.
    pub(crate) version: Option<String>,
    /// The document path inside the package, without an extension.
    pub(crate) document: String,
    /// The anchor, without its `#`.
    pub(crate) anchor: Option<String>,
}

/// Read a `spec://` address, or `None` when it is not one.
///
/// ```text
/// spec://org.vibevm.core/vibevm-docs/guide/start#install
/// spec://org.vibevm.core/vibevm-docs@0.1.0/guide/start
/// ```
pub(crate) fn parse(uri: &str) -> Option<Citation> {
    let rest = uri.strip_prefix("spec://")?;
    let (rest, anchor) = match rest.split_once('#') {
        Some((head, tail)) if !tail.is_empty() => (head, Some(tail.to_string())),
        Some((head, _)) => (head, None),
        None => (rest, None),
    };
    let mut segments = rest.split('/');
    let group = segments.next()?;
    let name = segments.next()?;
    let document: Vec<&str> = segments.collect();
    if group.is_empty() || name.is_empty() || document.is_empty() {
        return None;
    }
    if document
        .iter()
        .any(|part| part.is_empty() || *part == "." || *part == "..")
    {
        return None;
    }
    let (name, version) = match name.split_once('@') {
        Some((name, version)) if !name.is_empty() && !version.is_empty() => {
            (name, Some(version.to_string()))
        }
        Some(_) => return None,
        None => (name, None),
    };
    Some(Citation {
        coordinate: format!("{group}/{name}"),
        version,
        document: document.join("/"),
        anchor,
    })
}

/// Answer one `resolve` request.
pub(crate) fn answer(reader: &Reader, query: Option<&str>) -> Result<Response, ApiError> {
    let Some(uri) = parameter(query.unwrap_or(""), "uri") else {
        return Err(ApiError::bad_request(
            "`resolve` takes the citation to follow: `?uri=spec://<group>/<name>/<document>`",
        ));
    };
    let Some(citation) = parse(&uri) else {
        return Err(ApiError::bad_request(format!(
            "`{uri}` is not a `spec://` address — a citation names a group, a documentation \
             package and a document inside it"
        )));
    };

    // `prefix` is `<group>/<name>/<version>/`; a citation is followed
    // only inside the one package this reader was pointed at.
    let Some((coordinate, version)) = reader.prefix.trim_end_matches('/').rsplit_once('/') else {
        return Err(ApiError::internal("this reader has no coordinate"));
    };
    if citation.coordinate != coordinate {
        return Err(ApiError::not_found(format!(
            "this reader serves `{coordinate}` and the citation names `{}`",
            citation.coordinate
        )));
    }
    if let Some(asked) = &citation.version
        && asked != version
    {
        return Err(ApiError::not_found(format!(
            "this reader serves version {version} and the citation names {asked} — a reader \
             is pointed at one package, and a version is a different one"
        )));
    }

    let set = vibe_doc::pages::read_package(&reader.package_dir)
        .map_err(|e| ApiError::internal(e.to_string()))?;
    let wanted = format!("{}.xml", citation.document);
    let Some(page) = set.pages.iter().find(|page| page.rel == wanted) else {
        return Err(ApiError::not_found(format!(
            "this documentation carries no page `{}`",
            citation.document
        )));
    };

    let mut target = format!("{}{}{}/", reader.base, reader.prefix, citation.document);
    if let Some(anchor) = &citation.anchor
        && vibe_doc::manifest::page::anchors_of(&page.doc)
            .iter()
            .any(|held| held == anchor)
    {
        target.push('#');
        target.push_str(anchor);
    }
    Ok(super::found(&target))
}

/// One query parameter's value, percent-decoded.
///
/// Written here for the reason the path decoder is: the REFUSAL matters
/// more than the decoding, and a lenient reader would pass on something
/// it guessed at. A parameter that repeats answers with its first
/// spelling, which is what every server that does not want two answers
/// does.
fn parameter(query: &str, name: &str) -> Option<String> {
    for pair in query.split('&') {
        let Some((key, value)) = pair.split_once('=') else {
            continue;
        };
        if key != name {
            continue;
        }
        return super::decode(&value.replace('+', " "));
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_citation_is_read_into_its_four_parts() {
        assert_eq!(
            parse("spec://org.example/a-docs/guide/start#install"),
            Some(Citation {
                coordinate: "org.example/a-docs".to_string(),
                version: None,
                document: "guide/start".to_string(),
                anchor: Some("install".to_string()),
            })
        );
        assert_eq!(
            parse("spec://org.example/a-docs@0.1.0/start"),
            Some(Citation {
                coordinate: "org.example/a-docs".to_string(),
                version: Some("0.1.0".to_string()),
                document: "start".to_string(),
                anchor: None,
            })
        );
    }

    #[test]
    fn anything_that_is_not_a_citation_is_refused_rather_than_repaired() {
        for spelling in [
            "https://example.com/",
            "spec://org.example",
            "spec://org.example/a-docs",
            "spec://org.example/a-docs/",
            "spec://org.example/a-docs/../etc",
            "spec:///a-docs/start",
            "spec://org.example/@1.0/start",
        ] {
            assert!(parse(spelling).is_none(), "`{spelling}` must not resolve");
        }
    }

    #[test]
    fn a_parameter_is_decoded_and_the_first_spelling_wins() {
        assert_eq!(
            parameter("uri=spec%3A%2F%2Fa%2Fb%2Fc", "uri").as_deref(),
            Some("spec://a/b/c")
        );
        assert!(parameter("other=1", "uri").is_none());
    }
}
